tests/gradients/core/test_pulse_generator_gradient.py

```diff
@@ -0,0 +1,623 @@
+    )
+    def test_output_properties(self, num_wires, pardims, r_dtype, c_dtype):
+        """Test that the output for a skew-Hermitian input is real-valued
+        and has the correct shape(s)."""
+        import jax.numpy as jnp
+
+        r_dtype, c_dtype = getattr(jnp, r_dtype), getattr(jnp, c_dtype)
+        dim = 2**num_wires
+        gen_shapes = [pardim + (dim, dim) for pardim in pardims]
+        gens = tuple(np.random.random(sh) + 1j * np.random.random(sh) for sh in gen_shapes)
+        # make skew-Hermitian and cast to right complex dtype
+        gens = tuple((g - np.moveaxis(g.conj(), -2, -1)).astype(c_dtype) for g in gens)
+
+        coeffs = _one_parameter_paulirot_coeffs(gens, num_wires)
+        assert isinstance(coeffs, tuple)
+        assert len(coeffs) == len(pardims)
+        assert all(c.shape == (4**num_wires - 1, *pardim) for c, pardim in zip(coeffs, pardims))
+        assert all(c.dtype == r_dtype for c in coeffs)
+
+    @pytest.mark.parametrize("num_wires", [1, 2, 3])
+    def test_with_pauli_basis(self, num_wires):
+        """Test that the generators of all possible ``PauliRot`` gates are correctly decomposed
+        into canonical basis vectors as coefficients."""
+        # generators of PauliRot
+        paulirot_gens = -0.5j * pauli_basis_matrices(num_wires)
+        # With many entries, each containing one basis element
+        paulirot_coeffs = _one_parameter_paulirot_coeffs(paulirot_gens, num_wires)
+        assert qml.math.allclose(paulirot_coeffs, np.eye(4**num_wires - 1))
+        # With a "single entry" containing all basis elements
+        paulirot_coeffs = _one_parameter_paulirot_coeffs([paulirot_gens], num_wires)
+        assert qml.math.allclose(paulirot_coeffs, np.eye(4**num_wires - 1))
+
+
+@pytest.mark.jax
+class TestNonzeroCoeffsAndWords:
+    """Test the utility function _nonzero_coeffs_and_words."""
+
+    @pytest.mark.parametrize("num_wires", [1, 2, 3])
+    def test_all_zero(self, num_wires):
+        """Test that no coefficients and words are returned when all coefficients vanish."""
+        dim = 4**num_wires - 1
+        shapes = [(dim, 3), (dim,), (dim, 2, 5)]
+        # Generate zero coefficients only
+        coeffs = tuple(np.zeros(shape) for shape in shapes)
+        new_coeffs, words = _nonzero_coeffs_and_words(coeffs, num_wires)
+        # There should be as many coefficient sets as before, namely three, and they should
+        # be empty.
+        assert new_coeffs == [[], [], []]
+        # There should be an empty list of Pauli words left.
+        assert words == []
+
+    @pytest.mark.parametrize("num_wires", [1, 2, 3])
+    def test_separate_nonzero(self, num_wires):
+        """Test that a single coefficient in any of the coefficients is sufficient
+        to keep the Pauli word in the filter."""
+        # Create many coefficients, each just with one ``1`` entry at distinct places.
+        coeffs = tuple(np.eye(4**num_wires - 1))
```

Comment:
This way the absolute value of the coefficients is not monotone, which should not have an influence on the ordering of the Pauli words.
```suggestion
        # Create many coefficients, each greater or equal ``1`` at distinct places.
        rng = np.random.default_rng(42)
        coeffs = tuple(rng.uniform(1, 2, size=4**num_wires - 1))
```

Reply:
Thanks!